src/report/bulk.rs

```diff
@@ -0,0 +1,202 @@
+    /// How many of the payload bytes are populated with log data?
+    pub n_log_bytes: [u8; 4],
+    /// How many of the payload bytes are populated with extension data?
+    pub n_extension_bytes: [u8; 4],
+}
+const_assert_eq!(1, align_of::<WireBulkHeader>());
+const_assert_eq!(16, size_of::<WireBulkHeader>());
+
+/// Attempt to split a bulk report from its on-the-wire representation
+/// into its constituent parts, without unbounded copying or any allocation.
+///
+/// Returns the source location id, the log payload bytes, and the extension payload bytes.
+/// The log payload bytes are expected to be interepreted as little-endian `CompactLogItem`s.
+/// Payload alignment is not addressed.
+pub fn try_bulk_from_wire_bytes(
+    wire_bytes: &[u8],
+) -> Result<(TracerId, &[u8], ExtensionBytes), ParseBulkFromWireError> {
+    if wire_bytes.len() < size_of::<WireBulkHeader>() {
+        return Err(ParseBulkFromWireError::MissingHeader);
+    }
+    assert_eq_align!(u8, WireBulkHeader);
+    debug_assert_eq!(
+        0,
+        wire_bytes.as_ptr() as usize % align_of::<WireBulkHeader>()
+    );
+    let (header_bytes, payload_bytes) = wire_bytes.split_at(size_of::<WireBulkHeader>());
+    let wire_header = unsafe { &*(header_bytes.as_ptr() as *const WireBulkHeader) };
+    if wire_header.fingerprint != bulk_framing_fingerprint() {
+        return Err(ParseBulkFromWireError::InvalidFingerprint);
+    }
+    let raw_location_id = u32::from_le_bytes(wire_header.location_id);
+    let location_id = TracerId::new(raw_location_id)
+        .ok_or_else(|| ParseBulkFromWireError::InvalidTracerId(raw_location_id))?;
+    let n_log_bytes = u32::from_le_bytes(wire_header.n_log_bytes);
+    let n_extension_bytes = u32::from_le_bytes(wire_header.n_extension_bytes);
+
+    if payload_bytes.len() < n_log_bytes as usize {
+        return Err(ParseBulkFromWireError::IncompletePayload);
+    }
+    let (log_bytes, extension_bytes) = payload_bytes.split_at(n_log_bytes as usize);
+    if extension_bytes.len() < n_extension_bytes as usize {
+        return Err(ParseBulkFromWireError::IncompletePayload);
+    }
+    Ok((location_id, log_bytes, ExtensionBytes(extension_bytes)))
+}
+
+/// Everything that can go wrong when attempting to interpret a bulk report
+/// from the wire representation
+#[derive(Debug, PartialEq, Eq)]
+pub enum ParseBulkFromWireError {
+    /// The fingerprint didn't match expectations
+    InvalidFingerprint,
+    /// There weren't enough bytes for a full header
+    MissingHeader,
+    /// There weren't enough payload bytes (based on
+    /// expectations from inspecting the header).
+    IncompletePayload,
+    /// The data type was not one of the supported varieties
+    UnsupportedDataType(u8),
+    /// The tracer id didn't follow the rules for being
+    /// a valid ekotrace-location-specifying TracerId
+    InvalidTracerId(u32),
+}
+
+// TODO - pick a better fingerprint value
+const BULK_FRAMING_FINGERPRINT_SOURCE: u32 = 2_718_281_828;
```

Comment:
So, this is a magic number indicating 'bulk'? I'd expect to see it as a hex constant in the source, fwiw, since how we usually talk about protocol constants, and it makes it easy to grep for.

Reply:
Sure, will do.